metricbeat/mb/lightmodules.go

```diff
@@ -0,0 +1,288 @@
+// light metric set defaults
+func (m *LightMetricSet) Registration(r *Register) (MetricSetRegistration, error) {
+	registration, err := r.metricSetRegistration(m.Input.Module, m.Input.MetricSet)
+	if err != nil {
+		return registration, errors.Wrapf(err,
+			"failed to start light metricset '%s/%s' using '%s/%s' metricset as input",
+			m.Module, m.Name,
+			m.Input.Module, m.Input.MetricSet)
+	}
+
+	originalFactory := registration.Factory
+	registration.IsDefault = m.Default
+
+	// Light modules factory has to override defaults and reproduce builder
+	// functionality with the resulting configuration, it does:
+	// - Override defaults
+	// - Call module factory if registered (it wouldn't have been called
+	//   if light module is really a registered mixed module)
+	// - Call host parser if defined (it would have already been called
+	//   without the light module defaults)
+	// - Finally, call the original factory for the registered metricset
+	registration.Factory = func(base BaseMetricSet) (MetricSet, error) {
+		// Override default config on base module and metricset
+		base.name = m.Name
+		baseModule, err := m.baseModule(base.module)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create base module for light module")
+		}
+		base.module = baseModule
+
+		// Run module factory if registered, it will be called once per
+		// metricset, but it should be idempotent
+		moduleFactory := r.moduleFactory(m.Input.Module)
+		if moduleFactory != nil {
+			m, err := moduleFactory(*baseModule)
+			if err != nil {
+				return nil, err
+			}
+			base.module = m
+		}
+
+		// At this point host parser was already run, we need to run this again
+		// with the overriden defaults
+		if registration.HostParser != nil {
+			base.hostData, err = registration.HostParser(base.module, base.host)
+			if err != nil {
+				return nil, err
+			}
+			base.host = base.hostData.Host
+		}
+
+		return originalFactory(base)
+	}
+
+	return registration, nil
+}
+
+// baseModule does the configuration overrides in the base module configuration
+// taking into account the light metric set default configurations
+func (m *LightMetricSet) baseModule(from Module) (*BaseModule, error) {
+	baseModule := BaseModule{
+		name:   m.Module,
+		config: from.Config(),
+	}
+	var err error
+	// Set defaults
+	if baseModule.rawConfig, err = common.NewConfigFrom(m.Input.Defaults); err != nil {
+		return nil, errors.Wrap(err, "invalid input defaults")
+	}
+	// Copy values from user configuration
+	if err = from.UnpackConfig(baseModule.rawConfig); err != nil {
+		return nil, errors.Wrap(err, "failed to copy values from user configuration")
+	}
+	// Update module configuration
+	if err = baseModule.UnpackConfig(&baseModule.config); err != nil {
+		return nil, errors.Wrap(err, "failed to set module configuration")
+	}
+	return &baseModule, nil
+}
+
+func (r *LightModulesRegistry) readModule(moduleName string) (*LightModule, bool, error) {
```

Comment:
Improve the signature of this function. When you read it anywhere it looks like this `module, found, err := r.readModule(moduleName)` which is not telling you very much what is that `found` doing there that is not returned as an error. Also you can consider creating a `ModuleReader` with a `Exists() bool` and a `Read() (*LightModule, error)` and pass it to the `LightModulesRegistry` as a dependency. At the same time, this `*LightModule` maybe could be a `LightModule` interface, allowing other types of LightModules in the future instead of this concrete implementation.

As a help to improve this signature think about what it is to "read a module" (I don't know, because it has a too generic naming). Maybe "open module" will be more leading to do what it seems it's doing.